DOCSP-7798: CSFLE diagrams and other tweaks

ccho-mongodb opened the pull request:
JIRA: https://jira.mongodb.org/browse/DOCSP-7798

Staging: https://docs-mongodbcom-staging.corp.mongodb.com/ecosystem/docsworker/DOCSP-7798-csfle-diagrams/use-cases/client-side-field-level-encryption-guide.html

https://docs-mongodbcom-staging.corp.mongodb.com/ecosystem/docsworker/DOCSP-7798-csfle-diagrams/use-cases/client-side-field-level-encryption-local-key-to-kms.html

ccho-mongodb commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -24,13 +24,15 @@ information in the context of a real-world scenario:
 
-For a complete runnable example project, see the `Additional Information`_
-section.
+.. note::
```
This section was changed into a `note` based on user research that showed it was not visible enough:
This section was changed in response to feedback provided by

nlarew replied:
nit: Maybe this could be a note admonition titled **Download the Code**?

ccho-mongodb replied:
Sounds good to me, will update.

ccho-mongodb commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -250,15 +262,36 @@ information on the features of envelope encryption and key management
+and decrypt document fields, are stored in a key vault collection in the same
+MongoDB replica set as the encrypted data.
+
+.. admonition:: Local Key Provider is not suitable for production
```
Updated this section that mentioned the master key should be stored remotely based on confusion of user research subjects.

ccho-mongodb commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -158,12 +160,22 @@ compliant to data privacy regulations with MongoDB.
-This guide is meant for users new to Client-Side Field Level Encryption (CSFLE).
-It covers the why and how at a high level, and is not meant to produce working
-code in a single file.
+This section explains the following configuration and implementation
```
Prior text was confusing; provided outline of what was going to be implemented instead.

ccho-mongodb commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -986,28 +1028,40 @@ To view and download a runnable example, select your driver below:
       **GitHub:** `PyMongo CSFLE runnable example
       <https://github.com/mongodb-university/csfle-guides/python/tree/master/>`_
 
+Move to Production
```
Added this section to highlight the KMS guide.

ccho-mongodb commented on source/use-cases/client-side-field-level-encryption-local-key-to-kms.txt:
```diff
@@ -48,6 +48,13 @@ Convert to a Remote Master Key
+encryption keys** are created and stored when using a KMS provider for
+MongoDB CSFLE:
+
+.. image:: /figures/CSFLE_Create_Keys_KMS.png
```
Added diagram to the KMS guide.

nlarew replied:
Shouldn't the Client Application have to read the master key from AWS KMS similarly to how it does for the local key diagram?

![image](https://user-images.githubusercontent.com/15657698/70644582-4e2fd200-1c11-11ea-9a9e-93c198679431.png)
![image](https://user-images.githubusercontent.com/15657698/70644597-57b93a00-1c11-11ea-9cfa-5355ad7685ec.png)

ccho-mongodb replied:
Yes, good catch, will update the diagram. For accuracy, the data encryption key is just passed to the AWS KMS API by libmongocrypt which returns the encrypted data encryption key.

ccho-mongodb commented:
Reminder to self to compress diagrams in docs-assets once this PR is accepted.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -158,12 +160,22 @@ compliant to data privacy regulations with MongoDB.
+- Software required to run your client and server in your local development
+  environment.
+
+- Creation, setup, and validation of the encryption keys.
```
What does "setup" refer to here?

ccho-mongodb replied:
I think I meant configuration, and possibly storage. I'll make it more clear.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -158,12 +160,22 @@ compliant to data privacy regulations with MongoDB.
+- Configuration of the client for automatic field-level encryption.
+
+- Demonstration of the client writing, querying, and reading encrypted
+  fields.
```
nit: "Demonstration" is not a "configuration and implementation detail"

Maybe: "How to write, read, and query on encrypted fields"

ccho-mongodb replied:
Good catch, will update.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -158,12 +160,22 @@ compliant to data privacy regulations with MongoDB.
-To view and run minimal end-to-end examples based on this guide, refer to our
-`CSFLE guides <https://github.com/mongodb-university/csfle-guides>`_ repository.
+For complete runnable examples of CSFLE based on this scenario, see the
+`CSFLE guides <https://github.com/mongodb-university/csfle-guides>`_
```
In the new note above we refer readers to the **Download Example Project** section. Should we mirror that here or remove this entirely?

ccho-mongodb replied:
I voted to remove this, but I think @terakilobyte wanted to keep it. Any additional thoughts Nathan?

terakilobyte replied:
Let's remove and see how it performs with user feedback

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -250,15 +262,37 @@ information on the features of envelope encryption and key management
+keys** are created and stored in the MongoDB implementation of CSFLE:
+
+.. image:: /figures/CSFLE_Create_Keys_Local.png
+   :alt: Diagram that describes creating master and data encryption keys when using a local provider
```
This image does not render on the staged docs for me

![image](https://user-images.githubusercontent.com/15657698/70574570-6a306680-1b72-11ea-98b2-a9a719c271ce.png)

nlarew replied:
Upon further review, NONE of the images render for me

ccho-mongodb replied:
Will try to figure out how to make it work.

ccho-mongodb replied:
The images work now.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -250,15 +262,37 @@ information on the features of envelope encryption and key management
+
+   The Local Key Provider is an insecure method of storage and is therefore
+   **not recommended** if you plan to use CSFLE in production. Instead,
+   the master key should be managed in a :wikipedia:`Key Management
```
nits:
- active voice: "you should manage"
- we say "manage(d)" a lot here. Maybe we could re-org a bit to cut down? e.g.

```rst
Instead, use a KMS that stores the master key and remotely manages
storage and decryption for data keys.
```

ccho-mongodb replied:
Thanks, good catch, will update.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -250,15 +262,37 @@ information on the features of envelope encryption and key management
+   System <Key_management#Key_management_system>` (KMS) which manages key
+   storage and decryption of the data encryption keys remotely.
+
+   To learn more on using AWS KMS in your CSFLE implementation, read the
```
nits:
- "more on" -> "more about"
- Does the guide teach you _more about_ AWS KMS in CSFLE or does it teach you _how to use_ AWS KMS in CSFLE?

ccho-mongodb replied:
Thanks for spotting that; it's the latter, will update.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -858,11 +893,13 @@ JSON Schema. This operation creates a document similar to the following:
+   :alt: Diagram that shows the data flow for reading field-level encrypted data
+
+Queries on documents with encrypted fields can be run using conventional
+MongoDB driver methods. When a doctor performs a query in the MedcoMD
```
nit: active voice

You can query for documents with encrypted fields using conventional/standard MongoDB driver methods.

ccho-mongodb replied:
Thanks, will update.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -858,11 +893,13 @@ JSON Schema. This operation creates a document similar to the following:
+Queries on documents with encrypted fields can be run using conventional
+MongoDB driver methods. When a doctor performs a query in the MedcoMD
+management system to search for a patient by their SSN, the patient's data
+is returned unencrypted:
```
nit: Isn't this really that "the driver unencrypts the patient's data before returning it"?

ccho-mongodb replied:
Thanks, will update.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -858,11 +888,13 @@ JSON Schema. This operation creates a document similar to the following:
-for a patient with their SSN, the patient's data is returned
-unencrypted:
+.. image:: /figures/CSFLE_Read_Encrypted.png
+   :alt: Diagram that shows the data flow for reading field-level encrypted data
```
Don't read operations need to use the JSON schema to know which data key to decrypt with?

![image](https://user-images.githubusercontent.com/15657698/70644474-14f76200-1c11-11ea-90e0-ac272c1d7c85.png)

ccho-mongodb replied:
Yes, but the presence of the JSON Schema is represented by the top arrow "Validates query with automatic encryption configuration". We updated the doc to refer to the schema as "automatic encryption configuration (in JSON Schema format)" since it was a point of confusion in user research sessions.
Do you think it would be helpful to include a parenthetical (JSON Schema) under the current label? Otherwise, what would you suggest?

nlarew replied:
If we updated the term in the guide then this should make sense. The JSON schema parenthetical might help but isn't necessary.

ccho-mongodb commented:
@nlarew 

New diagrams https://github.com/mongodb/docs-assets/pull/39

I split up the diagrams for master key and data key generation to make each one more clear. Also added libmongocrypt and unencrypted keys as actors, although they do not invoke actions. This added some complexity, but fills in some of the details that I think you were requesting when we were deliberating over sequence diagrams.

Copy in the guides need to be updated to reference them and describe them in context.

nlarew commented on source/use-cases/client-side-field-level-encryption-guide.txt:
```diff
@@ -250,15 +257,42 @@ information on the features of envelope encryption and key management
+stored:
+
+.. image:: /figures/CSFLE_Data_Key_Local.png
+   :alt: Diagram that describes creating the data encryption key when using a locally-managed master key
```
Should this blurb/diagram about the data encryption keys go in the next step instead? i.e. **B. Create a Data Encryption Key**

ccho-mongodb replied:
I initially was thinking that it would be helpful to include in the same section so that I could explain what the master key would be used for, but it does make more sense to place the diagram in the next section.

nlarew commented on source/use-cases/client-side-field-level-encryption-local-key-to-kms.txt:
```diff
@@ -48,8 +47,17 @@ Convert to a Remote Master Key
+   :alt: Diagram that describes creating a master key when using a KMS provider
 
+.. image:: /figures/CSFLE_Data_Key_KMS.png
+   :alt: Diagram that describes creating a data encryption key when using a KMS provider
```
Love this updated diagram nice work!
